compiler/front/sexp_reporter.nim

```diff
@@ -0,0 +1,185 @@
+## Implementation of the structured CLI message generator. Using
+## `--msgFormat=sexp` will make compiler switch to the report hook
+## implemented in this module.
+
+import
+  experimental/[
+    sexp,
+    diff,
+    colortext,
+    sexp_diff
+  ],
+  ast/[
+    lineinfos,
+    ast,
+    reports
+  ],
+  front/[
+    options,
+    msgs
+  ],
+  std/[
+    strutils
+  ]
+
+
+var writeConf: ConfigRef
+
+import std/options as std_options
+
+proc addFields[T](s: var SexpNode, r: T, ignore: seq[string] = @[])
+
+proc sexp[T: object | tuple](obj: T): SexpNode =
+  result = newSList()
+  addFields(result, obj)
+
+proc sexp[T: object | tuple](obj: ref T): SexpNode =
+  result = newSList()
+  addFields(result, obj[])
+
+proc sexp*[E: enum](e: E): SexpNode = newSSymbol($e)
+
+proc sexpItems*[T](s: T): SexpNode =
```

Comment:
> Also, seriously this process shouldn't compile.

what process are you talking about and why it shouldn't compile?

Reply:
Typo, yeah it's strange that it's allowed by the compiler.

Reply:
> it's strange that it's allowed by the compiler.

again, **why**?

Reply:
Because a type `T` is not guaranteed to have an `items`, yes?

Reply:
It won't compile when I instantiate the proc like every other generic, so I guess this is not allowed? I just don't get why this remark is attached to a definition. sexp two lines above can have $ defined as {.error.} and it would've been an instantiation error as well